docs/admin/etcd.md

```diff
@@ -1,49 +1,163 @@
-This path can be prefixed by using the [kube-apiserver](/docs/admin/kube-apiserver) flag
-`--etcd-prefix="/foo"`.
+1. Detect which version of etcd we were previously running.
+   For that purpose, we have added a dedicated (metadata) file that will be
+   holding that information and will be stored in etcd-data-specific directory
+   (next to real etcd data). If the file doesn’t exist, we default it to 2.2.1
+   version.
+1. If we are in 2.2.1 version and are supposed to upgrade, backup
+   data.
+1. Based on the detected previous etcd version and the desired one (the desired
+   version is communicated via environment variable), do the upgrade steps (if
+   needed).
+   That means, for every minor etcd release between the detected one (excluding)
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not list on the default etcd
+   port.  They are hardcoded to listen on ports which the apiserver is not
+   configured to connect to, which means that apiserver won’t be able to connect
+   to those etcd's and no new data will be written in the meantime.
+1. If the desired api to use is v3 and detected on is v2, do the offline
+   migration from v2 to v3 data format.
+   For that we are using two tools:
+   1. ./etcdctl migrate - official tool for migration provided by CoreOS
+   1. custom script that is attaching TTLs to events in the etcd (etcdctl
+      migrate doesn’t support TTLs)
+1. After every successful step, update contents of “version file”.
+   This will protect us from the situation that e.g. something will crash in the
+   meantime and the “version file” will be completely unsynchronized with the
+   real data. Note that it is safe if the script will crash after the step is
+   done and before the file is updated - this will only result in redoing one
+   step in the next try.
 
-`etcd` is the only place that Kubernetes keeps state.
+All the previous steps are for the case where detected version is earlier or
+equal to the desired one. In the opposite case (i.e. requested rollback), the
+script:
 
-## Troubleshooting
+1. Checks if detected version is 3.0.x with v3 API, and desired is 2.2.1 with v2
+   API
+   We don’t support any other rollback.
+1. If so, we run the custom tool provided by etcd team to do the offline
+   rollback.  This tool reads the v3 formatted data and writes it back to disk
+   in v2 format.
+1. Finally update contents of “version file”.
 
-To test whether `etcd` is running correctly, you can try writing a value to a
-test key. On your master VM (or somewhere with firewalls configured such that
-you can talk to your cluster's etcd), try:
+### Upgrade procedure
+Simply modify the command-line in etcd manifest to:
 
-```shell
-curl -fs -X PUT "http://${host}:${port}/v2/keys/_test"
-```
+1. Run the migration script
```

Comment:
from https://github.com/kubernetes/kubernetes/issues/43366#issuecomment-287789004:

> [downgrading is] supported assuming that:
> 1. you are still using 'json' as the storage data format (even though the default is protobuf in 1.6)
> 2. you're not downgrading back to 2.2.1 image, but just using "3.0.17" image and setting ```ETCD_VERSION ``` to 2.2.1 

Documentation needs to make it crystal clear that you have to explicitly set `--storage-media-type=application/json` if you want to have the option to downgrade from etcd3 to etcd2.

It also needs to explain the "use the 3.0.17 image but run in 2.2.1 mode"... I'm not seeing that mentioned in this doc.